lib/core/waitForCoreChainLockSyncFactory.js

```diff
@@ -0,0 +1,70 @@
+const { ChainLock } = require('@dashevo/dashcore-lib');
+const ZMQClient = require('@dashevo/dashd-zmq');
+const waitForEvent = require('../util/waitForEvent');
+const ensureBlock = require('../util/ensureBlock');
+
+/**
+ * Check that core is synced (factory)
+ *
+ * @param {ZMQClient} coreZMQClient
+ * @param {RpcClient} coreRpcClient
+ * @param {LatestCoreChainLock} latestCoreChainLock
+ * @param {Logger} logger
+ *
+ * @returns {waitForCoreSync}
+ */
+function waitForCoreChainLockSyncFactory(
+  coreZMQClient,
+  coreRpcClient,
+  latestCoreChainLock,
+  logger,
+) {
+  /**
+   * Check that chainLock is synced
+   *
+   * @typedef waitForCoreChainLockSync
+   *
+   * @returns {Promise<LatestCoreChainLock>}
+   */
+  async function waitForCoreChainLockSync() {
+    await coreZMQClient.connect();
+    // listens to ChainLock zmq zmqpubrawchainlocksig event and updates the latestCoreChainLock.
+    // If we lost connection with Core (heartbeats with retry policy) we need to exit with fatal
+    // In case of regtest fallback we need to listen for new blocks using hashblock zmq event
+    coreZMQClient.subscribe(ZMQClient.TOPICS.rawchainlock);
+    logger.info('Subscribe to rawchainlock ZMQ room');
+
+    coreZMQClient.on(ZMQClient.TOPICS.rawchainlock, async (rawChainLockMessage) => {
+      const socketChainLock = new ChainLock(rawChainLockMessage);
+      logger.info(`Received new chainlock ${socketChainLock.toString()}`);
+      await ensureBlock(coreZMQClient, coreRpcClient, socketChainLock.blockHash);
+      latestCoreChainLock.update(socketChainLock);
+    });
+
+    // Because a Chainlock may happen before its block, we also subscribe to rawblock
+    coreZMQClient.subscribe(ZMQClient.TOPICS.hashblock);
+    logger.info('Subscribe to hashblock ZMQ room');
+
+    // We need to retrieve latest chainlock from our fully synced DashCore instance
+    try {
+      const rpcBestChainLockResponse = await coreRpcClient.getBestChainLock();
+      const chainlock = new ChainLock(rpcBestChainLockResponse.result);
```

Comment:
What if Chain Lock is not present?

Reply:
We actually need to handle the error case in order to provide a fallback way for syncing chainLock, else what we just got thrown out of there by the error (which is not what we want as the error is legit).